Make refresh a replicated action

brwe opened the pull request:
prerequisite to #9421
see also #12600

s1monw commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java:
```diff
@@ -0,0 +1,106 @@
+        logger.trace("{} flush request executed on replica", indexShard.shardId());
+    }
+    @Override
+    protected boolean resolveIndex() {
```
I know it's unrelated to this change but why don't we have defaults for these boolean they look the same in 90% fo the cases?

brwe replied:
I don't know. Change that in this pr or better in another one?

s1monw replied:
we can do another one

brwe replied:
opened https://github.com/elastic/elasticsearch/pull/13218

s1monw commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -1002,6 +1004,10 @@ private void doFinish() {
 
     }
 
+    protected boolean skipExecutionOnShadowReplicas() {
```
can you add a javadoc to this?

s1monw commented on core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java:
```diff
@@ -0,0 +1,143 @@
+
+    protected abstract ShardRequest newShardRequest(Request request, ShardId shardId);
+
+    private void finishAndNotifyListener(ActionListener listener, CopyOnWriteArrayList<ShardResponse> shardsResponses) {
```
can we maybe have unittests for this - this looks pretty unittestable?

brwe replied:
I added BroadcastReplicationTests. Let me know if this is what you meant.

s1monw commented on core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java:
```diff
@@ -0,0 +1,143 @@
+                    }
+                } else {
+                    final ShardId shardId = shardRouting.shardId();
+                    replicatedBroadcastShardAction.execute(newShardRequest(request, shardId), new ActionListener<ShardResponse>() {
```
if you factor `replicatedBroadcastShardAction.execute` and  ` GroupShardsIterator groupShardsIterator = clusterService.operationRouting().searchShards(clusterService.state(), indexNameExpressionResolver.concreteIndices(clusterService.state(), request), null, null);
` 
out of this you can get a nicely unit-testable class here having one method like `doInnerExecute(GroupSahrdsIterator iter,  final ActionListener<Response> listener)` ?

brwe replied:
I added BroadcastReplicationTests. Let me know if this is what you meant.

s1monw commented:
I love this change looks very good. I left some comments around unittesting

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Override
+    protected ShardIterator shards(ClusterState clusterState, InternalRequest request) {
+        return clusterService.operationRouting().shards(clusterService.state(), request.concreteIndex(), request.request().getShardId().id()).shardsIt();
```
can we use the incoming state here?

bleskes replied:
also, why not use indexShards() now that we see this as a write op - then we don't need to change the visibility of the shards methond on Operation Routing

brwe replied:
> can we use the incoming state here?

yes

> also, why not use indexShards() now that we see this as a write op

you mean `clusterService.operationRouting().indexShards()`? that one needs a type and id and we don't have that here. or is there another one that does not?

bleskes replied:
indeed the indexShards suggestion is bad - it is not the right construct here as it is tied to a single doc.  Since we are after shard ids here (not grouping) , I think we should simplify the API to return a list of shardIds (which will solve this too). See comments here : https://github.com/elastic/elasticsearch/pull/13068/files#r38203633

brwe replied:
This is now a method that returns List<ShardId>: https://github.com/elastic/elasticsearch/pull/13068/files#diff-8ec8c1c769c4acb6f880e4e15d2b96f6R120 Is that what your meant?

brwe replied:
scatch that, different shards() method...

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    protected ShardIterator shards(ClusterState clusterState, InternalRequest request) {
+        return clusterService.operationRouting().shards(clusterService.state(), request.concreteIndex(), request.request().getShardId().id()).shardsIt();
```
same comments as on flush...

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportRefreshAction.java:
```diff
@@ -19,100 +19,46 @@
-    protected ClusterBlockException checkGlobalBlock(ClusterState state, RefreshRequest request) {
-        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+    protected ReplicationRequest newShardRequest(RefreshRequest request, ShardId shardId) {
+        return new ReplicationRequest(request).setShardId(shardId);
```
we should reduce the timeout on the request to 0 (we don't wait to wait if things are not available) - also, can we make a test that make sure we get a quick response when there is no master/other block and or primary shard is not available?

brwe replied:
yes, test is here https://github.com/elastic/elasticsearch/pull/13068/files#diff-0fa8612b4126beeb59f608820b0bd1b6R206

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportFlushAction.java:
```diff
@@ -19,99 +19,45 @@
-    protected ClusterBlockException checkGlobalBlock(ClusterState state, FlushRequest request) {
-        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+    protected ShardFlushRequest newShardRequest(FlushRequest request, ShardId shardId) {
+        return new ShardFlushRequest(request).setShardId(shardId);
```
same comment as the refresh action, can we lower the timeout to 0 and test it returns quickly if we have a block/primary not available?

brwe replied:
yes, test is here https://github.com/elastic/elasticsearch/pull/13068/files#diff-0fa8612b4126beeb59f608820b0bd1b6R247

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java:
```diff
@@ -0,0 +1,145 @@
+
+    @Override
+    protected void doExecute(final Request request, final ActionListener<Response> listener) {
+        GroupShardsIterator groupShardsIterator = clusterService.operationRouting().searchShards(clusterService.state(), indexNameExpressionResolver.concreteIndices(clusterService.state(), request), null, null);
```
can we make this an abstract method and also use a ShardIterator? (the grouping doesn't really makes sense IMHO). Also I think we can just iterate on all shards for the given indices, no?

bleskes replied:
Skip my comment about ShardIterator, it inherits from ShardsIterator and thus needs ShardRouting instances (which are useless here). Instead, my just use a List<ShardId> as a return value for the new method. Also, thinking some more, maybe supply a default implementation that gives all shards for the indices back...

brwe replied:
This is now a method that returns List<ShardId>: https://github.com/elastic/elasticsearch/pull/13068/files#diff-8ec8c1c769c4acb6f880e4e15d2b96f6R120 Is that what your meant?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java:
```diff
@@ -0,0 +1,145 @@
+                        public void onResponse(ShardResponse shardResponse) {
+                            shardsResponses.add(shardResponse);
+                            if (responsesCountDown.countDown()) {
+                                logger.trace("replicated broadcast: got response from {}", shardId);
```
can we add the action name here? useful in this base classes which may have multiple implementations.

bleskes replied:
also, we should log this out of the countdown no?

brwe replied:
yes

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java:
```diff
@@ -0,0 +1,145 @@
+                        @Override
+                        public void onFailure(Throwable e) {
+                            logger.trace("replicated broadcast: got failure from {}", shardId);
+                            int totalNumCopies = clusterService.state().getMetaData().index(shardId.index().getName()).getNumberOfReplicas() + 1;
```
can we cache the cluster state we used when we started this operation? this way we know for sure the index is still in it.

bleskes replied:
note that we don't see shard not available exceptions as a failure - we typically ignore those and just let the successful count be different than the total count. We have to check the type of failure and protect for it.

brwe replied:
yes, I actually had test failures because of that. changed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java:
```diff
@@ -0,0 +1,145 @@
+    protected abstract ShardRequest newShardRequest(Request request, ShardId shardId);
+
+    private void finishAndNotifyListener(ActionListener listener, CopyOnWriteArrayList<ShardResponse> shardsResponses) {
+        logger.trace("replicated broadcast: got all shard responses");
```
can we please add the action here?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replicatedbroadcast/TransportBroadcastReplicationAction.java:
```diff
@@ -0,0 +1,145 @@
+        for (int i = 0; i < shardsResponses.size(); i++) {
+            ActionWriteResponse shardResponse = shardsResponses.get(i);
+            if (shardResponse == null) {
+                // non active shard, ignore
```
if we change the groupShardIterator to a ShardIterator - can this happen?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -37,10 +37,14 @@
+        return internalShardId;
+    }
+
     ShardId internalShardId;
```
this can be protected now, right?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -37,10 +37,14 @@
 
     public static final TimeValue DEFAULT_TIMEOUT = new TimeValue(1, TimeUnit.MINUTES);
 
+    public ShardId getShardId() {
```
can we mark this as nullable (and doc when it's null)? also, can we move it next to the setter, and make the naming consistent with the rest of this class? (i.e., shardId)

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -172,6 +172,7 @@ protected boolean ignoreReplicaException(Throwable e) {
         if (isConflictException(e)) {
             return true;
         }
+        // TODO should we check here for refresh and flush failures which should not make replica fail? if so, which are they?
```
I think it's OK to mark the shard as deleted - if there was some connectivity failure (i.e., the shard it self is not broken) *strictly* speaking the shard is no longer a good copy as it doesn't maintain the "all writes so far are visible to searches" semantics. I can see that it's potentially a big deal for nothing (after 1s it will be OK) but I think we should for now do the right thing and fail the shard?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -984,6 +986,14 @@ private void doFinish() {
     }
 
     /**
+     * Indicated whether this operation should be replicated to shadow replicas or not. If this method returns false the replication phase will be skipped.
```
I think the comment is flipped - if the method returns true , shadow replicas will be skipped. I think we will better off making this a positive - executeOnShadowReplicas?  we can also make this method say "shouldExecuteReplication" and have the default implementation do `return IndexMetaData.isIndexUsingShadowReplicas(indexMetaData.settings()) == false`. That will make the if where the method is being used simpler.

Also, we need to adapt the unit testing of this class for this (I don't think we have a test now for shadow replicas)

bleskes replied:
I forgot about that test. I think it's good for now.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -226,7 +226,7 @@ protected IndexShardRoutingTable shards(ClusterState clusterState, String index,
     }
 
-    protected IndexShardRoutingTable shards(ClusterState clusterState, String index, int shardId) {
+    public IndexShardRoutingTable shards(ClusterState clusterState, String index, int shardId) {
```
see comment above about this visibility change.

bleskes commented:
I agree with @s1monw that this looks great. I left minor comments here and there.

s1monw commented:
LGTM

brwe commented:
@s1monw thanks for the review! 

addressed all comments. @bleskes want to have one more look?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/BroadcastReplicationTests.java:
```diff
@@ -0,0 +1,289 @@
+
+    public BroadcastResponse executeAndAssertImmediateResponse(TransportBroadcastReplicationAction refreshAction, BroadcastRequest request) throws InterruptedException, ExecutionException {
+        Date beginDate = new Date();
+        BroadcastResponse response = (BroadcastResponse) refreshAction.execute(request).get();
```
you can use 

```
        BroadcastResponse response = (BroadcastResponse) refreshAction.execute(request).actionGet("30s");
```

which will throw an exception if things take longer than 30s.

Btw - if you need to measure duration , please use  System.nanoTime(), which doesn't suffer from ntp corrections.

brwe replied:
much nicer!

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/BroadcastReplicationTests.java:
```diff
@@ -0,0 +1,289 @@
+        assertFailure("all shards should fail with cluster block", executeAndAssertImmediateResponse(refreshAction, new RefreshRequest(index)), ClusterBlockException.class);
+    }
+
+    public BroadcastResponse executeAndAssertImmediateResponse(TransportBroadcastReplicationAction refreshAction, BroadcastRequest request) throws InterruptedException, ExecutionException {
```
naming - refreshAction?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/BroadcastReplicationTests.java:
```diff
@@ -0,0 +1,289 @@
+            actionWriteResponse.setShardInfo(new ActionWriteResponse.ShardInfo(1, 1, new ActionWriteResponse.ShardInfo.Failure[0]));
+            shardRequests.v2().onResponse(actionWriteResponse);
+        }
+        response.get();
```
nit: redundant...

bleskes commented:
Looks awesome. I only miss a test for the aggregation of results from multiple shard level responses.

brwe commented:
@bleskes addressed all comments and added test here: https://github.com/elastic/elasticsearch/pull/13068/files#diff-6030559b5ed4d55d9a754523f5c6ce6dR137

bleskes commented:
LGTM! (minor comments, no need for another review)
